### PR TITLE
Add markdown to notebook entries

### LINKED ISSUE #6060: [Notebooks] Formatted entries v2
**Is your feature request related to a problem? Please describe.**
Robust formatting in Notebook entries. Using Markdown: https://www.npmjs.com/package/snarkdown is a lightweight viable option.

**Describe the solution you'd like**
When editing a notebook entry, if you were to use MarkDown, when viewing the notebook entry, it would be formatted accordingly using HTML.

**Additional context**
First version is #6059

### PR DESCRIPTION
**Pull request by scottbell**
Closes #6060

### Describe your changes:

Add the [Marked](https://github.com/markedjs/marked) library, and consolidate the URL whitelisting/anchor tag generation.

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing document](https://github.com/nasa/openmct/blob/master/CONTRIBUTING.md)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/nasa/openmct/pulls) for the same update/change?
* [x] Is this change backwards compatible? For example, developers won't need to change how they are calling the API or how they've extended core plugins such as Tables or Plots.

### Author Checklist

* [x] Changes address original issue?
* [x] Tests included and/or updated with changes?
* [x] Command line build passes?
* [x] Has this been smoke tested?
* [x] Testing instructions included in associated issue OR is this a dependency/testcase change?

### Reviewer Checklist

* [ ] Changes appear to address issue?
* [ ] Reviewer has tested changes by following the provided instructions?
* [ ] Changes appear not to be breaking changes?
* [ ] Appropriate automated tests included?
* [ ] Code style and in-line documentation are appropriate?
* [ ] Has associated issue been labelled unverified? (only applicable if this PR closes the issue)
* [ ] Has associated issue been labelled bug? (only applicable if this PR is for a bug fix)

### CODE REVIEW

**Comment by scottbell** [comment]
At this point, I'll probably need some CSS help from @rukmini-bose or @charlesh88. We're rendering this sample input:

[sample.md](https://github.com/nasa/openmct/files/12725491/sample.md)

which renders the HTML:
```html
<div id="entry-37de3c3f-72e8-4603-a3d2-4bb69f0c1c73" class="c-ne__text c-ne__input" aria-label="Notebook Entry Input" tabindex="-1" contenteditable="true"><h1>h1 Heading</h1>
<h2>h2 Heading</h2>
<h3>h3 Heading</h3>
<h4>h4 Heading</h4>
<h5>h5 Heading</h5>
<h6>h6 Heading</h6>
<h2>Emphasis</h2>
<p><strong>This is bold text</strong></p>
<p><strong>This is bold text</strong></p>
<p><em>This is italic text</em></p>
<p><em>This is italic text</em></p>
<p><del>Strikethrough</del></p>
<h2>Blockquotes</h2>
<blockquote>
<p>Blockquote </p>
<blockquote>
<p>nested quote</p>
<blockquote>
<p>very nested</p>
</blockquote>
</blockquote>
</blockquote>
<h2>Lists</h2>
<p>Unordered</p>
<ul>
<li>apple</li>
<li>orange</li>
</ul>
<p>Ordered</p>
<ol>
<li>First item</li>
<li>Second item</li>
</ol>
<h2>Code Blocks</h2>
<p>Block code "fences"</p>
<pre><code>Sample text here...
</code></pre>
<h2>Tables</h2>
<table>
<thead>
<tr>
<th>Option</th>
<th>Description</th>
<th>Notes</th>
</tr>
</thead>
<tbody><tr>
<td>data</td>
<td>path to data files to supply the data that will be passed into templates.</td>
<td>first item</td>
</tr>
<tr>
<td>engine</td>
<td>engine to be used for processing templates. Handlebars is the default.</td>
<td>second item</td>
</tr>
<tr>
<td>ext</td>
<td>extension to be used for dest files.</td>
<td>third item</td>
</tr>
</tbody></table>
<h2>Links</h2>
<p>These need to be whitelisted:</p>
<p>link text</p>
<p>link with title</p>
</div>
```

This mostly works:
Screenshot 2023-09-26 at 11 10 30 AM
Screenshot 2023-09-26 at 11 10 37 AM

with the exception of:
1. lists - both unordered and ordered
2. italicized text
3. block quotes and code blocks might need a rectangle to distinguish?

**Comment by jvigliotta** [comment]
> @jvigliotta Here's what I've got so far.
> 
>  Screen.Recording.2023-09-25.at.4.53.44.PM.mov 
> There are few issues with it:
> 
> * CSS wise, it looks like we're not enabling certain things (e.g., `<em>`, and lists)
> * It looks like we're stripping out a lot of tags we'd like in markdown. I'll add them to the sanitize list.

This looks awesome Scott!

**Comment by charlesh88** [comment]
Have pushed up CSS changes that fix most of the problems you were seeing. May still do some sanding and shimming tomorrow, minor stuff like table widths and text contrast.

Also ran into a couple issues and a question:
1. Single line breaks need to work for regular and blockquoted text. We need to set Marked "breaks" == true in config?
1. Nested unordered (bullet) list items need to work. While "- Something" gets converted into a `<li>`, the renderer doesn't convert "  - Something" at all. Marked seems to support this out of the box, so not sure what's going on.
1. Leaving a space at the end of a list item before a double-return encodes a non-breaking space in a paragraph container between the two. The only way to fix it is to delete the trailing space at the end of the list item, which is really unintuitive. This sucks, can we do anything about this? 
1. Dragging an image from a random webpage didn't embed for me, not sure if that's supposed to work. Dragging from the desktop worked great though. Very nice!

[Exported JSON of my test Notebook used for screenshots below](https://github.com/nasa/openmct/files/12732787/6060.Markup.Notebook.json.zip)

**Marked up entry test, editing**
Screen Shot 2023-09-26 at 5 06 29 PM

**Marked up entry test, result**
Screen Shot 2023-09-26 at 5 06 04 PM

**Item 3, editing**
Screen Shot 2023-09-26 at 5 08 20 PM

**Item 3, result**
Screen Shot 2023-09-26 at 5 07 46 PM

**Comment by scottbell** [comment]
@charlesh88 

> Single line breaks need to work for regular and blockquoted text as noted above.
I think I've got that figured per your suggestion by adding `breaks: true` to the options. Here is the resultant HTML now:
```html
<p>Here's a line with a double-return before it.&nbsp;<br>Here's a line with a single return before it. It doesn't add a <code>&lt;br&gt;</code> before it, but should.</p>
```

**Comment by scottbell** [comment]
@charlesh88 
>Dragging an image from a random webpage didn't embed for me, not sure if that's supposed to work. Dragging from the desktop worked great though. Very nice!
Dragging from a webpage will only work if [CORS](https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS) is enabled for the website for the server hosting Open MCT. You should have at least gotten a nice notification saying "Failed to fetch image" and a more specific error in the console.

**Comment by scottbell** [comment]
@charlesh88 
>Nested unordered (bullet) list items need to work. While "- List item" gets converted into a <li>, the renderer doesn't convert " - Nested list item" at all. Marked seems to support this out of the box (it works on their demo site), so not sure what's going on.

I think this got fixed by the `breaks: true`. Here's the rendering with that enabled:
Screenshot 2023-09-27 at 9 25 35 AM

**Comment by scottbell** [comment]
@charlesh88
> Leaving a space at the end of a list item before a double-return encodes a non-breaking space in a paragraph container between the two. The only way to fix it is to delete the trailing space at the end of the list item, which is really unintuitive. This sucks, can we do anything about this?

This is a tough one to fix as Marked has already tokenized the extra space as separate from the list item, so we'd have to remove it either before being fed to Marked, or after. We could do this by either using regular expressions on the input text or resultant HTML, or parsing the output HTML into DOMs, manipulating it there, and then serializing the HTML back out again. Would it be OK to file this as a follow on issue?

**Comment by scottbell** [comment]
> @charlesh88
> 
> > Leaving a space at the end of a list item before a double-return encodes a non-breaking space in a paragraph container between the two. The only way to fix it is to delete the trailing space at the end of the list item, which is really unintuitive. This sucks, can we do anything about this?
> 
> This is a tough one to fix as Marked has already tokenized the extra space as separate from the list item, so we'd have to remove it either before being fed to Marked, or after. We could do this by either using regular expressions on the input text or resultant HTML, or parsing the output HTML into DOMs, manipulating it there, and then serializing the HTML back out again. Would it be OK to file this as a follow on issue?

I say that and the [Marked demo](https://marked.js.org/demo/?text=-%20Leaving%20a%20space%20at%20the%20end%20of%20a%20list%20item%20before%20a%20double-return%20encodes%20a%20non-breaking%20space%20in%20a%20paragraph%20container.%20The%20only%20way%20to%20fix%20it%20is%20to%20delete%20the%20trailing%20space%20at%20the%20end%20of%20list%20item%2C%20which%20is%20really%20unintuitive.%20This%20sucks%2C%20can%20we%20do%20anything%20about%20this%3F%20%0A%0ASee%3F&options=%7B%0A%20%22async%22%3A%20false%2C%0A%20%22breaks%22%3A%20false%2C%0A%20%22extensions%22%3A%20null%2C%0A%20%22gfm%22%3A%20true%2C%0A%20%22hooks%22%3A%20null%2C%0A%20%22pedantic%22%3A%20false%2C%0A%20%22silent%22%3A%20false%2C%0A%20%22tokenizer%22%3A%20null%2C%0A%20%22walkTokens%22%3A%20null%0A%7D&version=9.0.3) isn't doing this. Let me dig into this further.

**Comment by scottbell** [comment]
@charlesh88 and thank you very much for the CSS assistance - it look way better!

**Comment by charlesh88** [comment]
1. breaks: true was the fix, looks great.
2. Ok if we can't figure out the trailing space thing, but the biggest issue there is user frustration: you go to edit the entry to fix the extra space, delete what looks like 2 returns, save - and the extra space is back again.
3. Did get a "Failed to fetch image" message on browser window drag. That's a limited use case anyway, dragging from the desktop is probably much more valuable. All good.

**Comment by scottbell** [comment]
@charlesh88 For the trailing space and nested lists, I think they're both caused by having div with out a `whitespace` attribute on the CSS, thus when the content is loaded for editing, all the spaces are collapsed into a single space and the line breaks are tossed.

Setting `white-space: pre-wrap;` in the CSS I think is rendering (closer) to the HTML we want:
```scss
&__input {
    // Appended to __text element when Notebook is not in readOnly mode
    @include inlineInput;
    padding-left: $p;
    padding-right: $p;
    overflow: unset;
    margin-bottom: $interiorMargin;
    text-overflow: unset;
    white-space: pre-wrap;
  }
```

Here's the resultant trailing space HTML:
```html
<li>Leaving a space at the end of a list item before a double-return encodes a non-breaking space in a paragraph container. The only way to fix it is to delete the trailing space at the end of list item, which is really unintuitive. This sucks, can we do anything about this?</li>
</ul>
<p>See?</p>
```

and the nested lists:
```html
<ul>
    <li>List item<ul>
            <li>This nested list item isn't recognized but should be.</li>
        </ul>
    </li>
    <li>Another list item<ol>
            <li>Numbered item under an unordered list item&nbsp;isn't recognized but should be.</li>
            <li>Numbered item under an unordered list item&nbsp;isn't recognized but should be.</li>
        </ol>
    </li>
</ul>
```

Here are screenshots of the rendering, which still have a lot of space that we can probably compact:
Screenshot 2023-09-28 at 2 16 21 PM
Screenshot 2023-09-28 at 2 16 13 PM

Another alternative we could pursue is swapping out the `div` with a `textarea` a la Github and the Marked demo use as it preserves whitespace out of the box.

**Comment by scottbell** [comment]
@charlesh88 Here's the latest:

https://github.com/nasa/openmct/assets/9853862/9ba3a5e2-4eb7-4d89-a0e3-67285ee43a51

Had to change a lot of the way we were doing events as the `div` (or the `textarea`) doesn't always exist anymore. Also, on switch to the `textarea` for editing, we're adjust the size to accommodate the text. I couldn't find a nice way to do this in pure CSS. Let me know what you think!

**Comment by charlesh88** [comment]
Looking great! A couple notes:
- Sizing of the textarea based on its content is great, and yes, there is no way to do that with pure CSS.
- A somewhat important user interaction has gone away, and we need to bring back: the user should be required to select an entry, then click _again_ in the text area to edit it. This is to prevent inadvertent editing and to allow other actions related to selecting an entry, like tagging it, without putting the text into edit mode. Right now, clicking in the text area of an unselected entry immediately puts the text into edit mode.

**Comment by scottbell** [comment]
@charlesh88 i’ll add the two step focus/click again. I’ll also try to get the tests ready on Monday too so we can move this out of draft.

**Comment by charlesh88** [comment]
I just pushed a few changes for Shift Log, notably removed the hover styling for the entry text area when in a locked page. Apologize in advance if my linter messed up the code formatting in Notebook.vue, it was doing some kind of weird stuff.

**Comment by charlesh88** [comment]
We also need to handle entries in Search results. Right now, the result is showing the entirety of the entry, with markdown characters:

Screen Shot 2023-09-29 at 12 58 44 PM

Showing the whole entry is (suprpisingly) not a regression. This doesn't fly, particularly if folks will be doing longer entries now that they have markup. We need to do the following for Notebook entry search results:
1.  Do not include markdown characters.
2. Restrict to the first line of the entry, up to a max number of characters (64?) and add an ellipsis onto the end if more than one line and max chars are exceeded.
3. Just display the text without rich formatting, i.e. if the first line is an H1, don't display it that way in the results. If needed, this can be accomplished via CSS rules applied to the entry rather than stripping or suppressing html markup.

I

**Comment by scottbell** [comment]
@charlesh88 I think I've got your requests in:

https://github.com/nasa/openmct/assets/9853862/3298e2f3-8cb6-46bc-afac-b5d15558fe65

We're now:

- Two step editing. User must focus click, then click again on the entry to enable editing.
- Scrolling to the entry. I added this as the default scroll behavior was weird when we resized the box.
- For search results, we're parsing the markup, stripping all HTML tags (leaving only text), and possibly truncating to 64 characters. If we do truncate, we slap an ellipsis on the end.

**Comment by akhenry** [comment]
Before this is merged, I want to make sure we are not allowing any unsanitized HTML or JavaScript from the user?

**Review by akhenry — CHANGES_REQUESTED** [review]
See comments about sanitization

**Comment by scottbell** [comment]
> > See comments about sanitization
> 
> @akhenry We're sanitizing output from the Markdown [here](https://github.com/nasa/openmct/pull/7084/files#diff-84726fafd8b74e3a67bb63a563bd62a53363536291218e268fd6818b4f5932efR380). We're doing [the same thing](https://github.com/nasa/openmct/pull/7084/files#diff-ba025b05b602fc8a63d18ebe90e4bdb9e8b2ae50c6a677f0b1827416d0346bd1R148) before we render for annotation search results.

@akhenry Added [a further input sanitization](https://github.com/nasa/openmct/pull/7084/files#diff-84726fafd8b74e3a67bb63a563bd62a53363536291218e268fd6818b4f5932efR569), to prevent it from being saved in the first place.


https://github.com/nasa/openmct/assets/9853862/6ed4797f-b0e6-4a1f-9794-b7777a81c064

**Comment by akhenry** [comment]
@scottbell  Beautiful, thank you!

**Review by akhenry — APPROVED** [review]

File: src/ui/layout/search/AnnotationSearchResult.vue [hunk 1]
```diff
@@ -119,6 +105,9 @@ export default {
       return this.result.fullTagModels[0].foregroundColor;
     }
   },
+  beforeMount() {
```
Comment (ozyx) [inline src/ui/layout/search/AnnotationSearchResult.vue:108]: not sure what difference it makes, but this could be in `created()` as well

**Review by ozyx — APPROVED** [review]
Really awesome work! I tested all the usual Markdown syntax, including hyperlinks (which still respected the URL sanitization properly!) Looks solid.

**Comment by charlesh88** [comment]
@scottbell Fantastic work, was truly a pleasure working on this with you. Just ran the merged code locally and it's working great!

File: e2e/tests/functional/plugins/notebook/notebook.e2e.spec.js [hunk 2]
```diff
@@ -482,4 +482,42 @@ test.describe('Notebook entry tests', () => {
     expect.soft(await sanitizedLink.count()).toBe(1);
     expect(await unsanitizedLink.count()).toBe(0);
   });
+  test('Can add markdown to a notebook entry', async ({ page }) => {
```
[the remark this answers is not in the document]
Comment (scottbell) [reply e2e/tests/functional/plugins/notebook/notebook.e2e.spec.js:485]: I'll put in a ticket for that.

[the remark this answers is not in the document]
Comment (scottbell) [reply e2e/tests/functional/plugins/notebook/notebook.e2e.spec.js:485]: https://github.com/nasa/openmct/issues/7108

**Pull request merged**